### ISSUE TITLE
Consider major upgrades as information only upgrades

### ISSUE THREAD

**Issue by jeromectm** [sparkle-project/Sparkle#2070]
Hi,
I would like to avoid my paid users to overwrite their app with a major upgrade for which they will have to pay again. Is this possible?
I already use `sparkle:minimumAutoupdateVersion` to avoid automatic install, but I would like these major upgrades to be considered as information only upgrades. Maybe `sparkle:informationalUpdate` could be more flexible and allow to specify that **every version older than x.0** (and only those) should consider it as information only?
Or better, could the delegate decide if the user would have to pay, and change an `SUAppcastItem` as `isInformationOnlyUpdate`?

The only workaround I found is to implement `bestValidUpdateInAppcast:forUpdater:`, but:
- I need to call `[NSClassFromString(@"SUAppcastDriver") bestItemFromAppcastItems:[inAppcast items] comparator:[SUStandardVersionComparator defaultComparator]]` to get the default best item, and this is a private method
- `SUAppcastItem` is read-only, so I have to instantiate a new one from a modified copy of the original `SUAppcastItem`'s `propertiesDictionary`
- the new SUAppcastItem is no longer considered as `isMajorUpgrade`, which means the user can't ignore subsequent updates

Thanks!

**Comment by zorgiepoo** [comment]
I will first preface this post with we want to move towards solutions that allow the user to use Sparkle to update their app. Not being able to update in-app for a major upgrade is not a great user experience. It is a similar experience to using informational updates redirecting to the developer's website when the updater process gets hosed (and this is what I prefer informational only updates to be used for). If the user release notes are not sufficient in denoting a major upgrade, then improving this experience may involve us adding a UI hook so the app can inform the user or let them take the appropriate actions before Sparkle proceeds to updating the app.

With that said..

> I already use sparkle:minimumAutoupdateVersion to avoid automatic install, but I would like these major upgrades to be considered as information only upgrades. Maybe sparkle:informationalUpdate could be more flexible and allow to specify that every version older than x.0 (and only those) should consider it as information only?

If you use `sparkle:informationalUpdate` on `2.0` (for example) without specifying a version requirement, then any version `< 2.0` will see an informational only update. If you go with this approach you probably want to use a new appcast for `2.0` such that future updates like 2.0.1 will also not be seen by versions `< 2.0`. For any version `> 2.0` you no longer need or want an informational only update. If this doesn't cover all your cases then you'll need to elaborate on your situation more.

Please avoid using `-bestValidUpdateInAppcast:forUpdater:` if you can. Constructing new appcast items to mutate existing ones is especially misusing it.

**Comment by jeromectm** [comment]
> Not being able to update in-app for a major upgrade is not a great user experience

Agreed; however, replacing an app that the user has paid and which is 100% functional, with an app in demo mode, or which may have some disabled features, or which may expire after some time, that is in my opinion not a better user experience.

The user should be able to try the new release, and after some time decide if it is worth buying it and deleting the previous release, without the need to manually re-download the previous version from the developer's website if he decides not to upgrade.

Maybe Sparkle could handle that, and install an "appname 3.app" in the same folder as the renamed "appname 2.5.7.app" (when upgrading from 2.5.7 to 3.0.1 for example). And provide an API so the app can ask Sparkle to delete the old version and rename the new one (remove the version number) once the user has bought the upgrade.

Using a different appcast for minor updates of the new version is possible, however users who do not choose to buy version n.0 won't be notified for version n.1 or n.5, which may come with new features worth upgrading, nor he will be notified for version [n+1].0 (except if the developer puts version [n+1] in all appcasts). This is already what happens when the user skips a major upgrade (but I would prefer letting him chose if he wants to ignore this specific version, or all subsequent ones; and also, in my opinion, ignoring updates n.x should not ignore future major upgrades for versions [n+1].x), but not if he installs it to try, then manually reverts to the previous version.

Also, adding a delegate method to decide if an update is informational only (or should preserve the previous app version instead of replacing it) would allow the app to do this only if the user has actually bought the current version (as opposed to using it in demo mode or without any in-app purchase), and is not eligible to a free upgrade (if he bought it recently for example). And this would be very easy to implement both in Sparkle and for the developer :-)

Another solution would be to support downgrading to the last update preceding the major upgrade…

Thanks!

**Comment by zorgiepoo** [comment]
We don't allow downgrades for various reasons (including security and appcast selection) so I don't expect that to be supported.

We do have a "normalization" setting (at compile time, but it should probably be an Info.plist key) that allows renaming of an app. What you're suggesting for keeping the old version is similar to that except we should just not delete the old app version. I assume in your case the app bundle name you ship has changed too (if not you run into a similar problem when the user downloads/updates the app manually, thereby encouraging them to replace the old app). This might be feasible, however it would probably be up to the app to somehow communicate to the user that the old version is still installed and not replaced. Deleting the old version when the user wants it to be deleted is probably not feasible from Sparkle.

I would prefer to keep updates being informational as part of the feed than decided by the client. I wouldn't mind adding an upper bound version attribute there. Alternatively you can hardcode a list of versions to get "decent enough coverage" (although this is less than ideal, but a possibility today). This is somewhat at odds against the skip behavior as you noticed, but adding an option to skip only the specific version could be an option too (but this can get confusing, and may not bode well for apps that ship a lot of minor updates).

**Comment by zorgiepoo** [comment]
> Using a different appcast for minor updates of the new version is possible, however users who do not choose to buy version n.0 won't be notified for version n.1 or n.5, which may come with new features worth upgrading, nor he will be notified for version [n+1].0 (except if the developer puts version [n+1] in all appcasts). 

A way to address this with your current model is to post eg n.5 as an info only update to your old appcast. If it comes with new features worth upgrading you can bump the minimumAutoupdateVersion to n.5 and that way the user skipping [n.0, n.5) gets reset. This would also give you control such that n.1 for example is not a major bump worth notifying yet.

**Comment by zorgiepoo** [comment]
Sorry ignore the above post, I haven't looked at the logic in a while, it looks like it'll still skip versions upwards.

**Comment by zorgiepoo** [comment]
> I already use sparkle:minimumAutoupdateVersion to avoid automatic install, but I would like these major upgrades to be considered as information only upgrades. Maybe sparkle:informationalUpdate could be more flexible and allow to specify that every version older than x.0 (and only those) should consider it as information only?
Or better, could the delegate decide if the user would have to pay, and change an SUAppcastItem as isInformationOnlyUpdate?

I realize I am getting endlessly confused / forgetting Sparkle's behavior here so with anything I say here you should evaluate/test it. But to answer your original question:

* Use a single appcast
* Set <sparkle:minimumAutoupdateVersion> for all new updates >= major one.
* Set <sparkle:version> in <sparkle:informationalUpdate> to the latest minor sparkle version for new updates >= major one. Because of sparkle:minimumAutoupdateVersion above, users will upgrade to the latest minor sparkle version before being prompted of the major update. This effectively covers "every version older than x.0" already.

This doesn't address the "skip major updates" behavior but that is more involved.

**Comment by zorgiepoo** [comment]
> The only workaround I found is to implement bestValidUpdateInAppcast:forUpdater:, but:

Just for completeness sake, this is possible today to make updates selectively info-only like this, just that it's more of a hack than adding a new below-version attribute or similar. Check out `SUAppcastItem+Private.h` and `SPUAppcastItemStateResolver.h`, you can probably remove the `enclosure` in a newly created properties and pass that to create a new item. You won't get what would otherwise be the default selection, but you won't need it, just select the latest version using the standard version comparator. The appcast driver in 2.x filters out illegible updates before asking the delegate their choice and may ask twice (one for minor updates pass, another for major updates pass).

**Comment by jeromectm** [comment]
> Set sparkle:version in sparkle:informationalUpdate to the latest minor sparkle version for new updates >= major one. Because of sparkle:minimumAutoupdateVersion above, users will upgrade to the latest minor sparkle version before being prompted of the major update. This effectively covers "every version older than x.0" already.

Ah, yes, this will update to the latest minor update, then let the user manually download and install the major upgrade if he wants to. That is OK for me :-)

However, when the user skips a major upgrade, I would still prefer:
1. letting him choose if he wants to skip this version only, or all its minor updates as well
2. **not** letting him ignore subsequent major upgrades. It is not uncommon for my users to buy an upgrade from version n to n+2.

**Comment by zorgiepoo** [comment]
> Ah, yes, this will update to the latest minor update, then let the user manually download and install the major upgrade if he wants to. That is OK for me :-)

This is actually not perfect if the user skips the latest minor release, they will be on an old minor release and be prompted for new major update, I think. To make it perfectly sound, and upper bound requirement would still be needed.

2. I think is arguably a bug and a small change/fix to make. Maybe I had a reason not to do it but I can't think what it would be offhand.

1 or providing multiple skip choices is likely going to be complicated.

**Comment by zorgiepoo** [comment]
> Another related suggestion: when installing the latest minor update, temporarily ignoring a major upgrade, maybe `SULastCheckTime` could be left untouched, so the major upgrade will be proposed immediately after relaunching the app?

I don't understand the intent/desire behind this. I don't think I'd want to do this (as a user). It seems pretty custom and not a big difference either way (however adding in such a change in the framework may not be elegant). I think you can simulate it if you want via observing one of the delegate methods, checking if it's a minor update, and invoking checkForUpdatesInBackground on next relaunch if you want (if you know there is a major update).

--

I thought of the following changes that I think would meet both of our ends better:

* Change skipping minimumAutoupdateVersion n so it doesn't skip minimumAutoupdateVersion n + 1
* Remove alert/confirmation explaining to user the skip behavior when skipping major update
* Add a `<sparkle:version sparkle:below=..>` attribute to informational updates
* Add a `<sparkle:minimumAutoupdateVersion sparkle:below=..>` attribute to minimumAutoupdateVersion

UX would be simplified not needing to explain the user anything when skipping a major update or make promises on how skipping works. Major updates would be still be skipped upwards, however, developers can do something eg `<sparkle:minimumAutoupdateVersion sparkle:below="2.0">2.4</sparkle:minimumAutoupdateVersion>` such that 2.4 is notable enough to prompt users that have skipped 2.0 but not disable autoupdates for users > 2.0.

Let me know what you think about that.

**Comment by jeromectm** [comment]
Sorry, I am quite confused at this point. We are talking of multiple related things:
- disabling silent installing of major upgrades (prompt for user confirmation instead)
- automatically skipping subsequent updates (preferably minor updates only), if the user skips a major upgrade
- information-only updates when the current version is < n (in addition to always, or from specific version numbers only)

> Change skipping minimumAutoupdateVersion n so it doesn't skip minimumAutoupdateVersion n + 1

I really don't understand this one. It would have sense if you were talking of `SUSkippedMajorVersion` (and by "doesn't skip minimumAutoupdateVersion n + 1", I would understand: "doesn't skip appcast items with `sparkle:minimumAutoupdateVersion` > n")

> Remove alert/confirmation explaining to user the skip behavior when skipping major update

I guess I could then implement `updater:userDidSkipThisVersion:` and replace the `SUSkippedMajorVersion` preference by `SUSkippedMinorVersionKey` if I want (maybe after custom user confirmation) still notify minor updates?

> Add a <sparkle:version sparkle:below=..> attribute to informational updates (convenience, so additional appcast is not needed with enough foresight)

Good; however should it be an attribute, or an element inside `sparkle:informationalUpdate`, like the current `sparkle:version`? Having both an attribute and some elements is maybe a bit confusing.

> Add a <sparkle:minimumAutoupdateVersion sparkle:below=..> attribute to minimumAutoupdateVersion (convenience, so additional appcast is not needed with enough foresight)

Too confusing in my opinion. As this element is about disabling silent install, it is confusing to add an attribute about skipping notifications of minor / sub-major releases when the user has skipped a previous major upgrade. I think you even swapped 2.0 and 2.4 in your sample!

To recap, I would prefer that if the user skips a major upgrade, Sparkle also skips subsequent minor updates, but not subsequent major upgrades, without needing to configure this in the appcast. And ideally if I can choose to not skip subsequent minor updates from the delegate.

**Comment by zorgiepoo** [comment]
> I really don't understand this one. It would have sense if you were talking of SUSkippedMajorVersion (and by "doesn't skip minimumAutoupdateVersion n + 1", I would understand: "doesn't skip appcast items with sparkle:minimumAutoupdateVersion > n")

This is what you wanted, I might have worded it poorly. If you have `<sparkle:minimumAutoupdateVersion>2.0</sparkle:minimumAutoupdateVersion>` in one appcast item that the user skips, and there is `<sparkle:minimumAutoupdateVersion>3.0</sparkle:minimumAutoupdateVersion>` in a future appcast item, the latter is not skipped. It doesn't matter what these values actually are, Sparkle just cares that 3.0 > 2.0. They could even be 2.0 and 2.4 for example.

> I guess I could then implement updater:userDidSkipThisVersion: and replace the SUSkippedMajorVersion preference by SUSkippedMinorVersionKey if I want (maybe after custom user confirmation) still notify minor updates?

Too hacky, may break minor updates too. This should not be done.

> Good; however should it be an attribute, or an element inside sparkle:informationalUpdate, like the current sparkle:version? Having both an attribute and some elements is maybe a bit confusing.

I was thinking `<sparkle:informationalUpdate><sparkle:version sparkle:below>2.0</sparkle:version></sparkle:informationalUpdate>`

> Too confusing in my opinion. As this element is about disabling silent install, it is confusing to add an attribute about skipping notifications of minor / sub-major releases when the user has skipped a previous major upgrade. I think you even swapped 2.0 and 2.4 in your sample!

It is still about disabling silent install. `<sparkle:minimumAutoupdateVersion>2.4</sparkle:minimumAutoupdateVersion >` disables silent install for versions < 2.4.  `<sparkle:minimumAutoupdateVersion sparkle:below="2.0">2.4</sparkle:minimumAutoupdateVersion >` disables silent install only for users < 2.0. However you may be right and adding another element to "clear" the skip marker instead might be better. 

> To recap, I would prefer that if the user skips a major upgrade, Sparkle also skips subsequent minor updates, but not subsequent major upgrades, without needing to configure this in the appcast. 

We agree on this part.

> And ideally if I can choose to not skip subsequent minor updates from the delegate.

I would prefer this to be controllable and fine-grained from the feed.

**Comment by zorgiepoo** [comment]
I am now thinking of something like this:

```
<!-- A second minor version, but not worthy enough to notify skipped 1.x users now -->
<item>
	<sparkle:version>2.2</sparkle:version>
	<sparkle:minimumAutoupdateVersion>2.0</sparkle:minimumAutoupdateVersion>
	<sparkle:ignoreSkippedUpgradesBelow>2.1</sparkle:ignoreSkippedUpgradesBelow>
	<sparkle:informationalUpdates>
		<sparkle:version sparkle:below>2.0</sparkle:version>
	</sparkle:informationalUpdates>
</item>

<!-- Minor version that *might* have little more importance, worthy of notifying skipped 1.x users -->
<item>
	<sparkle:version>2.1</sparkle:version>
	<sparkle:minimumAutoupdateVersion>2.0</sparkle:minimumAutoupdateVersion>
	<sparkle:ignoreSkippedUpgradesBelow>2.1</sparkle:ignoreSkippedUpgradesBelow>
	<sparkle:informationalUpdates>
		<sparkle:version sparkle:below>2.0</sparkle:version>
	</sparkle:informationalUpdates>
</item>

<!-- Bugfix version of little importance -->
<item>
	<sparkle:version>2.0.1</sparkle:version>
	<sparkle:minimumAutoupdateVersion>2.0</sparkle:minimumAutoupdateVersion>
	<sparkle:informationalUpdates>
		<sparkle:version sparkle:below>2.0</sparkle:version>
	</sparkle:informationalUpdates>
</item>

<!-- Major version -->
<item>
	<sparkle:version>2.0</sparkle:version>
	<sparkle:minimumAutoupdateVersion>2.0</sparkle:minimumAutoupdateVersion>
	<sparkle:informationalUpdates>
		<sparkle:version sparkle:below>2.0</sparkle:version>
	</sparkle:informationalUpdates>
</item>

<!-- Preceding major version -->
<item>
	<sparkle:version>1.9</sparkle:version>
</item>
```

**Comment by jeromectm** [comment]
> I am now thinking of something like this (read starting from bottom to top):

Yes, this seems OK for me.
And a last item that will be notified to users who skipped v2, without needing to specify `sparkle:ignoreSkippedUpgradesBelow`:

```
<!-- A new major version, which will be notified (as information only) to any user -->
<item>
	<sparkle:version>3</sparkle:version>
	<sparkle:minimumAutoupdateVersion>3.0</sparkle:minimumAutoupdateVersion>
	<sparkle:informationalUpdates>
		<sparkle:version sparkle:below>3.0</sparkle:version>
	</sparkle:informationalUpdates>
</item>
```

**Comment by zorgiepoo** [comment]
- [x] Don't skip subsequent major upgrades (#2079)
- [x] Add `sparkle:belowVersion` element to informational updates (#2080)
- [ ] Add `sparkle:ignoreSkippedUgradesBelowVersion` element (almost have this done)
- [ ] Update generate_appcast for two new elements
- [ ] Update documentation

**Comment by zorgiepoo** [comment]
I don't have time to properly tag a new beta release right now. However, the changes for this have been made and integrated. Any improvements for installing an upgrade in-app without deleting the old version will need to be tracked separately.
